### PR TITLE
docs(golang-style): Design and apply rules for using "%s" and %q in fmt.Sprintf [AIR-3985]

### PR DESCRIPTION
**Pull request by host6**
```yaml
registered_at: 2026-05-20T10:26:13Z
change_id: 2605201026-fmt-sprintf-string-escape-rules
type: docs
baseline: e50729e8f343aceebe4189c5848147088e711df6
issue_url: https://untill.atlassian.net/browse/AIR-3985
scope: golang-style
archived_at: 2026-05-20T10:56:45Z
```
## Why
In many places the codebase uses fmt.Sprintf("%s", data) where data carries content that must be escaped before being embedded into a larger string (JSON, URL query, file path, shell argument, etc.). Without explicit guidance, developers and AI agents keep producing unsafe or ambiguous formatting that leads to injection-style defects and hard-to-diagnose bugs.

## What
A code style guide that prescribes the correct formatting verb and escaping helper for string data passed to fmt.Sprintf (and related formatting functions), a small utility package that makes JSON templates safe by default, and fixes to existing violations.

### Style guide
- Survey of current usages of fmt.Sprintf with %s/%q on string data across the repository
- Rules that distinguish when %q is sufficient, when %s must be combined with url.QueryEscape, when json.Marshal is required, and analogous cases for other embedding contexts (host:port, file paths, HTML, SQL, Go-quoted literals)
- Guidance packaged as a Claude/Augment skill (.claude/skills/golang-fmt-string-escape/SKILL.md) targeted at both human developers and AI coding agents
- Self-balanced {} requirement in fmt.Sprintf JSON templates

### New utility: pkg/goutils/jsonu

A drop-in fmt.Sprintf-style helper for building small JSON snippets safely:

- jsonu.Jprintf(format, args...) JSON-escapes string-like arguments: string, named ~string types, values implementing fmt.Stringer, and values implementing error (with Error() taking precedence over String(), mirroring fmt.Sprintf)
- Verb modes:
  - %s / %v emit JSON-escaped content without quotes (caller supplies the quotes in the template)
  - %q emits a complete JSON string literal (escaped content wrapped in double quotes)
- Stricter than fmt.Sprintf %q: emits JSON-valid escapes (\u000b instead of \v, \ufffd for invalid UTF-8) instead of Go-only escapes that JSON parsers reject
- Supports standard fmt flags, width and precision via fmt.Formatter
- README with before/after examples, full test suite (basic usage, edge cases, control chars, HTML chars, named string types, error/Stringer precedence, width padding)

### CODE REVIEW

File: pkg/btstrp/impl.go [hunk 1]
```diff
@@ -101,7 +101,7 @@ func deployAppPartitions(ctx context.Context, stage string, appparts appparts.IA
 
 func callDeployApp(federation federation.IFederation, sysToken string, app appparts.BuiltInApp) {
 	// Use Admin Endpoint to send requests
-	body := fmt.Sprintf(`{"args":{"AppQName":"%s","NumPartitions":%d,"NumAppWorkspaces":%d}}`, app.Name, app.NumParts, app.NumAppWorkspaces)
+	body := fmt.Sprintf(`{"args":{"AppQName":%q,"NumPartitions":%d,"NumAppWorkspaces":%d}}`, app.Name, app.NumParts, app.NumAppWorkspaces)
```
[the remark this answers is not in the document]
Comment (host6) [reply pkg/btstrp/impl.go:104]: wrong

File: pkg/processors/blobber/impl_write.go [hunk 2]
```diff
@@ -28,7 +28,7 @@ import (
 func getRegisterFunc(ctx context.Context, bw *blobWorkpiece) (err error) {
 	if bw.isPersistent() {
 		bw.registerFuncName = registerPersistentBLOBFuncQName
-		bw.registerFuncBody = fmt.Sprintf(`{"args":{"OwnerRecord":"%s","OwnerRecordField":"%s"}}`,
+		bw.registerFuncBody = fmt.Sprintf(`{"args":{"OwnerRecord":%q,"OwnerRecordField":%q}}`,
 			bw.blobMessageWrite.ownerRecord, bw.blobMessageWrite.ownerRecordField)
```
[the remark this answers is not in the document]
Comment (host6) [reply pkg/processors/blobber/impl_write.go:32]: wrong

File: pkg/router/impl_apiv2.go [hunk 3]
```diff
@@ -327,14 +327,14 @@ func requestHandlerV2_create_device(numsAppsWorkspaces map[appdef.AppQName]istru
 		login, pwd := coreutils.DeviceRandomLoginPwd()
 		pseudoWSID := coreutils.GetPseudoWSID(istructs.NullWSID, login, istructs.CurrentClusterID())
 		url := fmt.Sprintf("api/v2/apps/sys/registry/workspaces/%d/commands/registry.CreateLogin", pseudoWSID)
-		body := fmt.Sprintf(`{"args":{"Login":"%s","AppName":"%s","SubjectKind":%d,"WSKindInitializationData":"{}","ProfileCluster":%d},"unloggedArgs":{"Password":"%s"}}`,
+		body := fmt.Sprintf(`{"args":{"Login":%q,"AppName":%q,"SubjectKind":%d,"WSKindInitializationData":"{}","ProfileCluster":%d},"unloggedArgs":{"Password":%q}}`,
 			login, busRequest.AppQName, istructs.SubjectKind_Device, istructs.CurrentClusterID(), pwd)
```
[the remark this answers is not in the document]
Comment (host6) [reply pkg/router/impl_apiv2.go:331]: wrong

File: pkg/sys/workspace/impl.go [hunk 4]
```diff
@@ -383,7 +411,7 @@ func initializeWorkspaceProjector(time timeu.ITime, federation federation.IFeder
 			if wsDescr.AsInt64(Field_InitStartedAtMs) == 0 {
 				info("initStartedAtMs = 0. WS init was not started")
 				// WS[currentWS].c.sys.CUD(wsDescr.ID, initStartedAtMs)
-				body := fmt.Sprintf(`{"cuds": [{"sys.ID": %d,"fields": {"sys.QName": "%s","%s": %d}}]}`,
+				body := fmt.Sprintf(`{"cuds": [{"sys.ID": %d,"fields": {"sys.QName": %q,%q: %d}}]}`,
 					wsDescr.ID(), appdef.QNameCDocWorkspaceDescriptor, Field_InitStartedAtMs, time.Now().UnixMilli())
```
[the remark this answers is not in the document]
Comment (host6) [reply pkg/sys/workspace/impl.go:388]: wrong

File: pkg/router/impl_reply_v1.go [hunk 5]
```diff
@@ -152,7 +152,11 @@ func reply_v1(requestCtx context.Context, w http.ResponseWriter, responseCh <-ch
 			sendSuccess = writeResponse(w, jsonErr)
 		} else {
 			writeHeaderOnce()
-			sendSuccess = writeResponse(w, fmt.Sprintf(`"status":%d,"errorDescription":"%s"`, http.StatusInternalServerError, *responseErr))
+			errFragment := jsonu.Jprintf(`"status":%d,"errorDescription":%q`, http.StatusInternalServerError, (*responseErr).Error())
+			if elemsCount == 0 {
```
Comment (maxim-ge) [inline pkg/router/impl_reply_v1.go:156]: Why we added this check here?

**Review by maxim-ge — CHANGES_REQUESTED** [review]
See comments

File: pkg/router/impl_reply_v1.go (hunk 5 above)
Comment (host6) [reply pkg/router/impl_reply_v1.go:156]: Query is requested, no sections sent yet, non-SysError occured. Then need to take the error in {} to get the correct json.
Actually this branch is not covered by tests, but the check is added correctly

**Review by maxim-ge — APPROVED** [review]
